opencog/query/DefaultPatternMatchCB.cc

```diff
@@ -389,33 +391,34 @@ bool DefaultPatternMatchCB::is_self_ground(const Handle& ptrn,
                                            const Handle& grnd,
                                            const HandleMap& term_gnds,
                                            const OrderedHandleSet& varset,
-                                           int quote_level)
+                                           Quotation quotation)
 {
 	Type ptype = ptrn->getType();
 
 	// Unwrap quotations, so that they can be compared properly.
-	if (ptype == QUOTE_LINK or ptype == UNQUOTE_LINK)
+	if (Quotation::is_quotation_type(ptype))
 	{
 		// Wow, if we are here, and patern==grnd, this must be
-		// a self-grounding, as I don't beleive there is any other
+		// a self-grounding, as I don't believe there is any other
 		// valid way to get to here.
-		if (ptype == QUOTE_LINK and 0 == quote_level and ptrn == grnd) return true;
-		if (ptype == UNQUOTE_LINK and 1 == quote_level and ptrn == grnd) return true;
-
-		if (ptype == QUOTE_LINK) quote_level++;
-		else quote_level--;
+		if (quotation.consumable(ptype) and ptrn == grnd) return true;
 
 		const Handle& qpat = ptrn->getOutgoingAtom(0);
-		return is_self_ground(qpat, grnd, term_gnds, varset, quote_level);
```

Comment:
OK, yes, I see now.  It seems weird to me that the quote-unwrapping happens in the PTM ctor, but perhaps this is dictated by local-quote, and perhaps we've spent enough time on this already.